### PR TITLE
Upgrade to Python 3.11

### PR DESCRIPTION
**Pull request by timotta**

### CODE REVIEW

File: tests/signals/handlers/test_rabbitmq.py [hunk 1]
```diff
@@ -45,14 +46,14 @@ async def test_startup_initializes_and_starts_one_consumer_per_route(
         )
         app = App(connections=[connection])
         app.routes_registry = self.routes_registry
-        app.loop = Mock(create_task=CoroutineMock())
+        app.loop = Mock(create_task=AsyncMock())
         # asynctest.MagicMock(
```
Comment (leotok) [inline tests/signals/handlers/test_rabbitmq.py:50]: @timotta O que acha de remover esse cometário ou então adaptá-lo pra não usar o `asynctest`?

File: tests/rabbitmq/test_rabbitmq_consumer.py [hunk 2]
```diff
@@ -633,15 +636,19 @@ async def test_start_always_calls_sleep(self):
...
-        ), asynctest.patch.object(
+        ), patch.object(
             asyncio, "sleep"
-        ) as sleep_mock, asynctest.patch.object(
+        ) as sleep_mock, patch.object(
             consumer, "clock_task", side_effect=[True, True]
         ):
-            queue_mock = CoroutineMock(
-                consume=CoroutineMock(), connect=CoroutineMock()
+            queue_mock = Mock(
```
Comment (leotok) [inline tests/rabbitmq/test_rabbitmq_consumer.py:646]: @timotta Nesse `queue_mock` está sendo usada a classe `Mock`, mas antes era `CoroutineMock` e na linha 667 quando é criado outro `queue_mock é usada a classe `AsyncMock`. Está certo ou era para ser o `AsyncMock` aqui tb?

Comment (leotok) [reply tests/rabbitmq/test_rabbitmq_consumer.py:646]: Mesma dúvida na linha 691 e 713

File: tests/signals/handlers/test_rabbitmq.py (hunk 1 above)
Comment (timotta) [reply tests/signals/handlers/test_rabbitmq.py:50]: @leotok eu tentei mexer o minimo possivel (sem remover os comentarios ja existentes) mas acho que seria uma boa pratica se os owners @daltonmatos e @diogommartins acharem por bem eu dou uma limpa também.

File: tests/rabbitmq/test_rabbitmq_consumer.py (hunk 2 above)
Comment (timotta) [reply tests/rabbitmq/test_rabbitmq_consumer.py:646]: @leotok estava errado antes queue_mock é apenas um objeto (e o CoroutineMock ao contrário do MagicMock não propaga a magica). Não causava dano usar CoroutineMock, mas também não trazia beneficio e causava confusão (vc esperaria em algum momento um await queue_mock())

File: asyncworker/time.py [hunk 3]
```diff
@@ -31,6 +31,7 @@ def __init__(self, seconds: Union[float, int]) -> None:
         self._tick_event = asyncio.Event()
         self._running: Optional[bool] = None
         self._main_task: Optional[asyncio.Future] = None
+        self._sleep = asyncio.sleep
```
Comment (diogommartins) [inline asyncworker/time.py:34]: Não precisamos guardar uma referência pro `asyncio.sleep` no objeto. Basta fazer o `patch` direto no `asyncio.sleep`, @timotta:

```patch
Index: tests/test_time.py

<+>UTF-8
===================================================================
diff --git a/tests/test_time.py b/tests/test_time.py
--- a/tests/test_time.py	(revision 1f7bf2bbe7a0175493a1fcd90ca5419c39a55af2)
+++ b/tests/test_time.py	(date 1677466264709)
@@ -55,10 +55,11 @@
                 event.clear.assert_not_called()
                 await clock.stop()
 
-            clock._sleep = AsyncMock(side_effect=tick_check)
-
-            clock._running = True
-            await self.loop.create_task(clock._run())
-            event.clear.assert_called_once()
+            with patch(
+                "asyncworker.time.asyncio.sleep", AsyncMock(side_effect=tick_check)
+            ) as sleep:
+                clock._running = True
+                await self.loop.create_task(clock._run())
+                event.clear.assert_called_once()
 
-            clock._sleep.assert_awaited_once_with(2)
+                sleep.assert_awaited_once_with(2)
Index: asyncworker/time.py

<+>UTF-8
===================================================================
diff --git a/asyncworker/time.py b/asyncworker/time.py
--- a/asyncworker/time.py	(revision 1f7bf2bbe7a0175493a1fcd90ca5419c39a55af2)
+++ b/asyncworker/time.py	(date 1677466160996)
@@ -31,7 +31,6 @@
         self._tick_event = asyncio.Event()
         self._running: Optional[bool] = None
         self._main_task: Optional[asyncio.Future] = None
-        self._sleep = asyncio.sleep
 
     def __aiter__(self) -> AsyncIterator:
         if self._running is not None:
@@ -55,7 +54,7 @@
     async def _run(self) -> None:
         while self._running:
             self._tick_event.set()
-            await self._sleep(self.seconds)
+            await asyncio.sleep(self.seconds)
             self._tick_event.clear()
 
     async def stop(self) -> None:

```

**Review by diogommartins — CHANGES_REQUESTED** [review]

File: .github/workflows/pull-request.yaml [hunk 4]
```diff
@@ -10,7 +10,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        python: [3.7, 3.8, 3.9, 3.x]
+        python: [3.9, 3.10, 3.11]
```
Comment (diogommartins) [inline .github/workflows/pull-request.yaml:13]:
```suggestion
        python: ["3.9", "3.10", "3.11"]
```

File: .github/workflows/main.yaml [hunk 5]
```diff
@@ -12,7 +12,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        python: [3.7, 3.8, 3.9, 3.x]
+        python: [3.9, 3.10, 3.11]
```
Comment (diogommartins) [inline .github/workflows/main.yaml:15]:
```suggestion
        python: ['3.9', '3.10', '3.11']
```

File: .github/workflows/main.yaml [hunk 6]
```diff
@@ -12,7 +12,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        python: [3.7, 3.8, 3.9, 3.x]
+        python: ['3.9', '3.10', '3.11']
```
Comment (diogommartins) [inline .github/workflows/main.yaml:15]: Aqui removemos o `3.x`. Não é uma boa voltar com ele pra que a gente eventualmente rode (e saiba se algo quebrou) no `3.12`, `3.13`, ... ?

**Comment by diogommartins** [comment]
## 3.9.16
```
Traceback (most recent call last):
[20](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507635808#step:8:21)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/bin/py.test", line 5, in <module>
[21](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507635808#step:8:22)
    from pytest import console_main
[22](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507635808#step:8:23)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/pytest/__init__.py", line 5, in <module>
[23](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507635808#step:8:24)
    from _pytest._code import ExceptionInfo
[24](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507635808#step:8:25)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/_pytest/_code/__init__.py", line 2, in <module>
[25](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507635808#step:8:26)
    from .code import Code
[26](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507635808#step:8:27)
  File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/_pytest/_code/code.py", line 60, in <module>
[27](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507635808#step:8:28)
    from exceptiongroup import BaseExceptionGroup
[28](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507635808#step:8:29)
ModuleNotFoundError: No module named 'exceptiongroup'
```
https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507635808

## 3.10.10

```
Traceback (most recent call last):
[20](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507636022#step:8:21)
  File "/opt/hostedtoolcache/Python/3.10.10/x64/bin/py.test", line 5, in <module>
[21](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507636022#step:8:22)
    from pytest import console_main
[22](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507636022#step:8:23)
  File "/opt/hostedtoolcache/Python/3.10.10/x64/lib/python3.10/site-packages/pytest/__init__.py", line 5, in <module>
[23](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507636022#step:8:24)
    from _pytest._code import ExceptionInfo
[24](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507636022#step:8:25)
  File "/opt/hostedtoolcache/Python/3.10.10/x64/lib/python3.10/site-packages/_pytest/_code/__init__.py", line 2, in <module>
[25](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507636022#step:8:26)
    from .code import Code
[26](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507636022#step:8:27)
  File "/opt/hostedtoolcache/Python/3.10.10/x64/lib/python3.10/site-packages/_pytest/_code/code.py", line 60, in <module>
[27](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507636022#step:8:28)
    from exceptiongroup import BaseExceptionGroup
[28](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507636022#step:8:29)
ModuleNotFoundError: No module named 'exceptiongroup'
[29](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507636022#step:8:30)
Error: Process completed with exit code 1.
```
https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507636022

Você teve algum problema para executar os testes localmente nessas versões, @timotta ? Vi que o 3.11 passou pelos testes de boa, mas [quebrou no upload de coverage](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507636228)

**Comment by diogommartins** [comment]
> ## 3.9.16
> ```
> Traceback (most recent call last):
> [20](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507635808#step:8:21)
>   File "/opt/hostedtoolcache/Python/3.9.16/x64/bin/py.test", line 5, in <module>
> [21](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507635808#step:8:22)
>     from pytest import console_main
> [22](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507635808#step:8:23)
>   File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/pytest/__init__.py", line 5, in <module>
> [23](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507635808#step:8:24)
>     from _pytest._code import ExceptionInfo
> [24](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507635808#step:8:25)
>   File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/_pytest/_code/__init__.py", line 2, in <module>
> [25](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507635808#step:8:26)
>     from .code import Code
> [26](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507635808#step:8:27)
>   File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/site-packages/_pytest/_code/code.py", line 60, in <module>
> [27](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507635808#step:8:28)
>     from exceptiongroup import BaseExceptionGroup
> [28](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507635808#step:8:29)
> ModuleNotFoundError: No module named 'exceptiongroup'
> ```
> 
> https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507635808
> 
> ## 3.10.10
> ```
> Traceback (most recent call last):
> [20](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507636022#step:8:21)
>   File "/opt/hostedtoolcache/Python/3.10.10/x64/bin/py.test", line 5, in <module>
> [21](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507636022#step:8:22)
>     from pytest import console_main
> [22](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507636022#step:8:23)
>   File "/opt/hostedtoolcache/Python/3.10.10/x64/lib/python3.10/site-packages/pytest/__init__.py", line 5, in <module>
> [23](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507636022#step:8:24)
>     from _pytest._code import ExceptionInfo
> [24](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507636022#step:8:25)
>   File "/opt/hostedtoolcache/Python/3.10.10/x64/lib/python3.10/site-packages/_pytest/_code/__init__.py", line 2, in <module>
> [25](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507636022#step:8:26)
>     from .code import Code
> [26](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507636022#step:8:27)
>   File "/opt/hostedtoolcache/Python/3.10.10/x64/lib/python3.10/site-packages/_pytest/_code/code.py", line 60, in <module>
> [27](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507636022#step:8:28)
>     from exceptiongroup import BaseExceptionGroup
> [28](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507636022#step:8:29)
> ModuleNotFoundError: No module named 'exceptiongroup'
> [29](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507636022#step:8:30)
> Error: Process completed with exit code 1.
> ```
> 
> https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507636022
> 
> Você teve algum problema para executar os testes localmente nessas versões, @timotta ? Vi que o 3.11 passou pelos testes de boa, mas [quebrou no upload de coverage](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507636228)
> 
> ```
> Error: you must supply a CC_TEST_REPORTER_ID ENV variable or pass it via the -r flag
> ```

**Comment by diogommartins** [comment]
Existe um secret CC_TEST_REPORTER_ID no projeto. Talvez só esteja faltando definir uma env com esse secret?

File: .github/workflows/pull-request.yaml [hunk 7]
```diff
@@ -38,13 +38,13 @@ jobs:
         run: pipenv run all-tests
 
       - name: Coverage upload
-        if: ${{matrix.python == '3.8'}}
+        if: ${{matrix.python == '3.11'}}
         uses: paambaati/codeclimate-action@v2.7.5
         env:
           CC_TEST_REPORTER_ID: ${{secrets.CC_TEST_REPORTER_ID}}
```
Comment (diogommartins) [inline .github/workflows/pull-request.yaml:44]: Env sendo definida aqui

Comment (diogommartins) [reply .github/workflows/pull-request.yaml:44]: @daltonmatos pedi uma permissão na org `async-worker` relacionada ao codecov. Deve ter chegado alguma coisa no seu e-mail. Pode autorizar ?

Comment (daltonmatos) [reply .github/workflows/pull-request.yaml:44]: Aprovei. Estamos adicionando codecov no repo do async-worker? A ideia seria trocar o codeclimate por ele?

**Comment by daltonmatos** [comment]
Pessoal, 

estou acompanhando um pouco de longe esse PR mas desde já agradeço muito o esforço e assim que tiver mais tempo vou olhar com mais detalhe e responder/opinar as dúvidas. 

Valeu!!

File: asyncworker/time.py (hunk 3 above)
Comment (timotta) [reply asyncworker/time.py:34]: @diogommartins vc tem razão, é mania de ex-javeiro de preferir fazer as coisas por injeção de dependência. Farei aqui a modificação

File: .github/workflows/main.yaml (hunk 6 above)
Comment (timotta) [reply .github/workflows/main.yaml:15]: Faz sentido! Na boa, acho que tirei o 3.x por distração, não foi consciente.

Comment (timotta) [reply .github/workflows/main.yaml:15]: > 
> Você teve algum problema para executar os testes localmente nessas versões, @timotta ? Vi que o 3.11 passou pelos testes de boa, mas [quebrou no upload de coverage](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507636228)
> 
> ```
> Error: you must supply a CC_TEST_REPORTER_ID ENV variable or pass it via the -r flag
> ```

Localmente testei apenas na versão 3.11. Vou testar agora nessas versões e ajustar para que fiquem OK.

**Comment by timotta** [comment]
> 
> Você teve algum problema para executar os testes localmente nessas versões, @timotta ? Vi que o 3.11 passou pelos testes de boa, mas [quebrou no upload de coverage](https://github.com/async-worker/async-worker/actions/runs/4305340383/jobs/7507636228)
> 
> ```
> Error: you must supply a CC_TEST_REPORTER_ID ENV variable or pass it via the -r flag
> ```

Localmente testei apenas na versão 3.11. Vou testar agora nessas versões e ajustar para que fiquem OK.

**Comment by timotta** [comment]
> > ModuleNotFoundError: No module named 'exceptiongroup'

Parece que esse erro estava presente no pytest 7.2.1. Atualizado para 7.2.2 o problema resolveu no python 3.9, 3.10 e 3.11 se manteve.

File: .github/workflows/pull-request.yaml (hunk 7 above)
Comment (diogommartins) [reply .github/workflows/pull-request.yaml:44]: Tem algo esquisito na relação entre o [codeclimate e PRs de forks](https://github.com/codeclimate/test-reporter/issues/452). Me parece que ele não tá passando env nesse contexto. Com as ultimas alterações do @timotta, [só a step do codeclimate continua quebrando](https://github.com/async-worker/async-worker/actions/runs/4331343146/jobs/7573751123) :/

**Comment by timotta** [comment]
@daltonmatos ainda estamos com erro na execução do upload de cobertura pela falta da env `CC_TEST_REPORTER_ID`. Olhando o log da action da pra ver que a env realmente está vazia. Me pergunto se isso não está acontecendo por alguma proteção de segurança (por ser um PR via fork, faz sentido ter uma proteção para que quem faz o PR não tenha como obter as secrets)

File: .github/workflows/pull-request.yaml (hunk 7 above)
Comment (timotta) [reply .github/workflows/pull-request.yaml:44]: Acho que faz sentido a secret não ficar exposta nesse PR pois é um PR vindo de um fork. Se essa secret ficasse exposta a qualquer PR via fork, seria facil obter ela adicionando actions para fazer upload dela pra um server do dono do fork.

Comment (diogommartins) [reply .github/workflows/pull-request.yaml:44]: Exato. Por padrão, secrets não são compartilhados com forks, [apesar dessa opção existir](https://github.blog/2020-08-03-github-actions-improvements-for-fork-and-pull-request-workflows/). Se não queremos compartilhar, não bastaria alterar a condição para não rodar em forks, algo como ` if: env.CC_TEST_REPORTER_ID && matrix.python_version == '3.11'`  @timotta ?

File: .github/workflows/pull-request.yaml [hunk 8]
```diff
@@ -38,13 +38,13 @@ jobs:
         run: pipenv run all-tests
 
       - name: Coverage upload
-        if: ${{matrix.python == '3.8'}}
+        if: ${{matrix.python == '3.11'}}
```
Comment (diogommartins) [inline .github/workflows/pull-request.yaml:41]:
```suggestion
        if: env.CC_TEST_REPORTER_ID && matrix.python_version == '3.11'
```

File: .github/workflows/pull-request.yaml (hunk 7 above)
Comment (timotta) [reply .github/workflows/pull-request.yaml:44]: @diogommartins comitei a sugestão do @daltonmatos que é exata a essa que vc sugeriu aí

**Review by diogommartins — APPROVED** [review]

**Comment by daltonmatos** [comment]
Acho podemos sim @timotta. O que fizemos aqui, apesar de ser apenas update de compatibilidade com python, melhor gerarmos uma [release breaking](https://async-worker.github.io/async-worker/versions.html#atualizacao-de-dependencias).

Vou publicar o draft da versão `0.20.0`. 

Lembro do @diogommartins ter comentado sobre um problema nos workflows que apesar do python escolhido ser uma versão X o workflow estava ainda rodando com uma versão Y. Conseguimos descobrir o que estava causando isso?

**Comment by daltonmatos** [comment]
@timotta 

PR para podermos fechar a release `0.20.0`. https://github.com/async-worker/async-worker/pull/30

**Comment by daltonmatos** [comment]
Feito. https://pypi.org/project/async-worker/0.20.0/

**Pull request merged**